src/os/bluestore/BlueStore.h

```diff
@@ -1638,6 +1641,11 @@ class BlueStore : public ObjectStore,
     pool_opts_t pool_opts;
     ContextQueue *commit_queue;
 
+    // Deferred write dependencies. Buffers of deferred writes than get cloned must be marked as clean
+    // on all cloned Onodes after the main deferred txc is completed.
+    std::mutex deferred_seq_dependencies_lock;
+    std::map<uint64_t, std::set<OnodeRef>> deferred_seq_dependencies;
```

Comment:
What happens with this collection when we split collection/cache. Doesn't it need splitting too?

Reply:
Buffers are addressed by onode and the onode is move to its respective collection so I think this is ok.

Reply:
Perhaps I'm missing something but what's about the following scenario:
1. Write to Onode1 (aka O1) is being deferred. It belongs to Collection1(aka C1).
2.  O1 is cloned to O2. C1 gets O2 in deferred_seq_dependencies.
3. C1 is split into C1_1 and C1_2. O1 is moved into C1_2 while deferred_seq_dependencies content is left in C1_1.
4. finish_write is called on O1 and it uses deferred_seq_dependencies from O1->c (=C1_2) to notify O2. But it's not there.

Reply:
I Deferred IO must be completed before splitting cache as mentioned in this line:
```c++

int BlueStore::_split_collection(TransContext *txc,
				CollectionRef& c,
				CollectionRef& d,
				unsigned bits, int rem)
{
  dout(15) << __func__ << " " << c->cid << " to " << d->cid << " "
	   << " bits " << bits << dendl;
  std::unique_lock l(c->lock);
  std::unique_lock l2(d->lock);
  int r;

  // flush all previous deferred writes on this sequencer.  this is a bit
  // heavyweight, but we need to make sure all deferred writes complete
  // before we split as the new collection's sequencer may need to order
  // this after those writes, and we don't bother with the complexity of
  // moving those TransContexts over to the new osr.
  _osr_drain_preceding(txc);

```

To be fair, I don't like how these semantics are hard to parse mentally.